flow/flow.go

```diff
@@ -0,0 +1,459 @@
+package flow
+
+import (
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/gobuffalo/pop/v5"
+
+	"github.com/ory/hydra/client"
+	"github.com/ory/hydra/consent"
+	"github.com/ory/hydra/x"
+	"github.com/ory/x/sqlcon"
+	"github.com/ory/x/sqlxx"
+)
+
+// FlowState* constants enumerate the states of a flow. The below graph
+// describes possible flow state transitions.
+//
+//graph TD
+//    LOGIN_INITIALIZED --> LOGIN_UNUSED
+//    LOGIN_UNUSED --> LOGIN_USED
+//    LOGIN_UNUSED --> LOGIN_ERROR
+//    LOGIN_USED --> CONSENT_INITIALIZED
+//    CONSENT_INITIALIZED --> CONSENT_UNUSED
+//    CONSENT_UNUSED --> CONSENT_UNUSED
+//    CONSENT_UNUSED --> CONSENT_USED
+//    CONSENT_UNUSED --> CONSENT_ERROR
+const (
+	// FlowStateLoginInitialized applies before the login app either
+	// accepts or rejects the login request.
+	FlowStateLoginInitialized = int16(1)
+
+	// FlowStateLoginUnused indicates that the login has been authenticated, but
+	// the User Agent hasn't picked up the result yet.
+	FlowStateLoginUnused = int16(2)
+
+	// FlowStateLoginUsed indicates that the User Agent is requesting consent and
+	// Hydra has invalidated the login request. This is a short-lived state
+	// because the transition to FlowStateConsentInitialized should happen while
+	// handling the request that triggered the transition to FlowStateLoginUsed.
+	FlowStateLoginUsed = int16(3)
+
+	// FlowStateConsentInitialized applies while Hydra waits for a consent request
+	// to be accepted or rejected.
+	FlowStateConsentInitialized = int16(4)
+
+	FlowStateConsentUnused = int16(5)
+	FlowStateConsentUsed   = int16(6)
+	FlowStateLoginError    = int16(128)
+	FlowStateConsentError  = int16(129)
```

Comment:
It appears these two are unused in the code. Should we keep them?

Reply:
In order to benefit from the error states (rather than adding yet another thing that needs to be checked when working with errors), we'd need to change a lot of the error handling logic. But this is PR is about refactoring the database, so I decided to leave it as a TODO. Should I remove the error states completely and just add a note that they could be introduced in the future?

Reply:
Adding a note could be really helpful for code readers :)